### DISCUSSION TITLE
yup's `oneOf` needs to be optional, but when undefined, errors field is empty but formState is invalid

### DISCUSSION THREAD

**Discussion by deannashih** [react-hook-form/react-hook-form#9075] [Q&A] [answered]
Hi! Weird situation here...

I have a field within my validation schema (using `yupResolver`) that should be one of several values but should also be optional.

Pared down, it looks like:
```
field: yup
      .string()
      .oneOf(['value1', 'value2', 'value3'])
      .when(
        'anotherField',
        (anotherField: boolean, schema: yup.StringSchema) => {
          if (!anotherField) {
            return schema.required();
          }
          return schema;
        }
      )
```
Basically, it should only be required depending on the value of `anotherField`. This may be a yup question, but it was working previously with Formik, and I've been working to convert Formik over to React Hook Form so I figured I'd start here.

Initial values for both `field` and `anotherField` are empty strings. When I set `anotherField` to `true`, `field` remains hidden and is therefore still unset (as expected.)

However, once everything else is valid, in the parent component where I have:

```
  const { formState: { isValid, errors } } = useForm({
    defaultValues: initialValues,
    resolver: yupResolver(validationSchema),
    mode: 'onTouched',
  });
```

`isValid` is false, while `errors` is empty.

I've tried transforming the `field` value to `undefined` when it's an empty string (considering `field` in typescript is set as optional), to no avail.

Basically, not sure why the `formState` is not valid, while there are no errors. I've pinned it down to the `oneOf` needing to be required when it shouldn't be. One more thing to note: once I toggle the values of `anotherField` and have it go back to `true` (i.e., set it to true --> false --> true), the form becomes valid, but I'm trying to get it to validate immediately.

Not sure if I should be asking here or over at `yup`. Stackoverflow is in read-only mode under maintenance atm :(

**Comment by leapful** [comment, marked as answer]
@deannashih Could you please share a codesandbox that can re-produce your problem ?

**Reply by deannashih** [reply]
@leapful thanks for responding!

I ended up figuring out the issue, which was twofold.

Regardless, [here's the codesandbox](https://codesandbox.io/s/dazzling-payne-ie7ej6?file=/src/Form.tsx) of a very contrived, pared down example of the much more nested and convoluted issue I was facing.

1) The solution to allowing `''` while keeping the original `.oneOf` values in tact was to add a `transform` check, rendering the form valid if the field remained untouched under certain conditions. It was loosely related to [another problem I was facing](https://github.com/jquense/yup/issues/971#issuecomment-961402894), where the solution was to basically allow the empty string as a valid value by making it undefined. ¯\_(ツ)_/¯

`.transform((value) => (value === "" ? undefined : value))`

What ends up happening is it strips out the field altogether on submit, which was fine for my use case. The main thing the `transform` fixed was that it allowed to form to proceed as valid, as expected.

2) Then there was the tangential issue I was having with the value of `anotherField` not triggering validation immediately either. I was using a custom wrapper component that had `react-select`'s `Select` under the hood. The underlying issue is found in [this answer](https://stackoverflow.com/a/68349375/5940856):

> ...`Select` doesn't expose input's ref. Therefore, you need to wrap `Select` with [`Controller`](https://react-hook-form.com/get-started) to register the component."

Editing my wrapper component directly was not a viable solution (there are a lot of Formik components that still use it), but the explanation was enough to let me know that was the reason validation wasn't being fired on change. My solution was to update `onChange` to manually `trigger` each field using this component.

i.e.
```
    onChange={(e) => {
              onChange(e);
              trigger("anotherField");
            }}
```
instead of
```
  onChange={onChange}
```

Did not realize validation between formik and react-hook-form would be suuuuch a trip lol. Posting in the hopes that others may find this useful down the line.

**Reply by leapful** [reply]
@deannashih :

Thanks for sharing your solution with codesandbox.

Formik has a pre-processing form values before passing to 3rd party schema validation library.

https://github.com/jaredpalmer/formik/blob/e677bea8181f40e6762fc7e7fb009122384500c6/packages/formik/src/Formik.tsx#L1081-L1104

**Reply by deannashih** [reply]
@leapful ahh amazing, thank you for posting this! a TIL/nuance about Formik I didn't know I needed to know — but this definitely explains a lot of my pain points of having to deal with the transition from Formik --> RHF. 😅

**Reply by bluebill1049** [reply]
thanks, @leapful that's awesome, you even dig into the source code. 🐰
